File: plugins/capawesome-capacitor-bluetooth-low-energy/android/src/main/java/io/capawesome/capacitorjs/plugins/bluetoothle/classes/results/GetServicesResult.java
package io.capawesome.capacitorjs.plugins.bluetoothle.classes.results;

import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import androidx.annotation.NonNull;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import io.capawesome.capacitorjs.plugins.bluetoothle.BluetoothLowEnergyHelper;
import io.capawesome.capacitorjs.plugins.bluetoothle.interfaces.Result;
import java.util.List;
import org.json.JSONArray;

public class GetServicesResult implements Result {

    private List<BluetoothGattService> services;

    public GetServicesResult(List<BluetoothGattService> services) {
        this.services = services;
    }

    @NonNull
    public JSObject toJSObject() {
        JSArray servicesResult = new JSArray();
        for (BluetoothGattService service : services) {
            JSArray characteristicsResult = new JSArray();
            for (BluetoothGattCharacteristic characteristic : service.getCharacteristics()) {
                JSArray descriptorsResult = new JSArray();
                for (BluetoothGattDescriptor descriptor : characteristic.getDescriptors()) {
                    JSObject descriptorResult = new JSObject();
                    descriptorResult.put("id", descriptor.getUuid().toString());
                    descriptorsResult.put(descriptorResult);
                }

                JSObject characteristicResult = new JSObject();
                characteristicResult.put("id", characteristic.getUuid().toString());
                characteristicResult.put("descriptors", descriptorsResult);
                characteristicResult.put("properties", createCharacteristicPropertiesResult(characteristic));
                characteristicsResult.put(characteristicResult);
            }

            JSObject serviceResult = new JSObject();
            serviceResult.put("id", service.getUuid().toString());
            serviceResult.put("characteristics", characteristicsResult);
            servicesResult.put(serviceResult);
        }

        JSObject result = new JSObject();
        result.put("services", servicesResult);
        return result;
    }

    @NonNull
    private JSObject createCharacteristicPropertiesResult(BluetoothGattCharacteristic characteristic) {
        JSObject properties = new JSObject();
        properties.put("broadcast", (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_BROADCAST) > 0);
        properties.put("read", (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_READ) > 0);
        properties.put(
            "writeWithoutResponse",
            (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE) > 0
        );
        properties.put("write", (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_WRITE) > 0);
        properties.put("notify", (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_NOTIFY) > 0);
        properties.put("indicate", (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_INDICATE) > 0);
        properties.put(
            "authenticatedSignedWrites",
            (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_SIGNED_WRITE) > 0
        );
        properties.put("extendedProperties", (characteristic.getProperties() & BluetoothGattCharacteristic.PROPERTY_EXTENDED_PROPS) > 0);
        return properties;
    }
}

File: plugins/capawesome-capacitor-bluetooth-low-energy/ios/Plugin/Classes/Results/GetServicesResult.swift
import Foundation
import CoreBluetooth
import Capacitor

@objc public class GetServicesResult: NSObject, Result {
    let services: [CBService]

    init(services: [CBService]) {
        self.services = services
    }

    public func toJSObject() -> AnyObject {
        var servicesResult = JSArray()
        for service in services {
            var characteristicsResult = JSArray()
            for characteristic in (service.characteristics ?? []) {
                var descriptorsResult = JSArray()
                for descriptor in (characteristic.descriptors ?? []) {
                    var descriptorResult = JSObject()
                    descriptorResult["id"] = BluetoothLowEnergyHelper.convertCBUUIDToString(uuid: descriptor.uuid)
                    descriptorsResult.append(descriptorResult)
                }

                var characteristicResult = JSObject()
                characteristicResult["id"] = BluetoothLowEnergyHelper.convertCBUUIDToString(uuid: characteristic.uuid)
                characteristicResult["descriptors"] = descriptorsResult
                characteristicResult["properties"] = createCharacteristicPropertiesResult(characteristic: characteristic)
                characteristicsResult.append(characteristicResult)
            }

            var serviceResult = JSObject()
            serviceResult["id"] = BluetoothLowEnergyHelper.convertCBUUIDToString(uuid: service.uuid)
            serviceResult["characteristics"] = characteristicsResult
            servicesResult.append(serviceResult)
        }

        var result = JSObject()
        result["services"] = servicesResult
        return result as AnyObject
    }

    private func createCharacteristicPropertiesResult(characteristic: CBCharacteristic) -> JSObject {
        var propertiesResult = JSObject()
        propertiesResult["broadcast"] = characteristic.properties.contains(.broadcast)
        propertiesResult["read"] = characteristic.properties.contains(.read)
        propertiesResult["writeWithoutResponse"] = characteristic.properties.contains(.writeWithoutResponse)
        propertiesResult["write"] = characteristic.properties.contains(.write)
        propertiesResult["notify"] = characteristic.properties.contains(.notify)
        propertiesResult["indicate"] = characteristic.properties.contains(.indicate)
        propertiesResult["authenticatedSignedWrites"] = characteristic.properties.contains(.authenticatedSignedWrites)
        propertiesResult["extendedProperties"] = characteristic.properties.contains(.extendedProperties)
        propertiesResult["notifyEncryptionRequired"] = characteristic.properties.contains(.notifyEncryptionRequired)
        propertiesResult["indicateEncryptionRequired"] = characteristic.properties.contains(.indicateEncryptionRequired)
        return propertiesResult
    }
}
